homeassistant/components/atome/sensor.py

```diff
@@ -0,0 +1,319 @@
+"""Linky Atome."""
+import logging
+from datetime import timedelta
+
+import voluptuous as vol
+from pyatome.client import AtomeClient
+from pyatome.client import PyAtomeError
+
+from homeassistant.const import (
+    CONF_PASSWORD,
+    CONF_USERNAME,
+    CONF_NAME,
+    DEVICE_CLASS_POWER,
+    POWER_WATT,
+    ENERGY_KILO_WATT_HOUR,
+)
+from homeassistant.components.sensor import PLATFORM_SCHEMA
+from homeassistant.helpers.entity import Entity
+from homeassistant.util import Throttle
+import homeassistant.helpers.config_validation as cv
+
+_LOGGER = logging.getLogger(__name__)
+
+DEFAULT_NAME = "atome"
+
+SESSION_RENEW_INTERVAL = timedelta(minutes=55)
+
+LIVE_SCAN_INTERVAL = timedelta(seconds=30)
+DAILY_SCAN_INTERVAL = timedelta(seconds=150)
+WEEKLY_SCAN_INTERVAL = timedelta(hours=1)
+MONTHLY_SCAN_INTERVAL = timedelta(hours=1)
+YEARLY_SCAN_INTERVAL = timedelta(days=1)
+
+LIVE_NAME = "Atome Live Power"
+DAILY_NAME = "Atome Daily"
+WEEKLY_NAME = "Atome Weekly"
+MONTHLY_NAME = "Atome Monthly"
+YEARLY_NAME = "Atome Yearly"
+
+LIVE_TYPE = "live"
+DAILY_TYPE = "day"
+WEEKLY_TYPE = "week"
+MONTHLY_TYPE = "month"
+YEARLY_TYPE = "year"
+
+ICON = "mdi:flash"
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend(
+    {
+        vol.Required(CONF_USERNAME): cv.string,
+        vol.Required(CONF_PASSWORD): cv.string,
+        vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
+    }
+)
+
+
+def setup_platform(hass, config, add_entities, discovery_info=None):
+    """Set up the Atome sensor."""
+    username = config[CONF_USERNAME]
+    password = config[CONF_PASSWORD]
+
+    data = AtomeData(username, password)
+
+    @Throttle(LIVE_SCAN_INTERVAL)
+    def update_live():
+        """Update the live power usage."""
+        data.get_live_power()
```

Comment:
Put the throttle on the method in the data class instead and have the entity call that method. Then we don't need the extra update functions here.

Pass the data method name to call instead to the entities and use `getattr`.

If we name the data methods consistently we could use the sensor type name together with `getattr`.

```py
getattr(self._data, f"update_{self._sensor_type}_usage")
```

Reply:
I took inspiration from the neurio_sensor component, which behaves like that. Thought I would do the same as behaviors are very similars

Reply:
I think I made what you want, what do you think ?
I also think there may be a better way to write the part in the sensor update function. What do you think ?